specs/phase0/weak-subjectivity.md

```diff
@@ -59,38 +77,81 @@ a safety margin of at least `1/3 - SAFETY_DECAY/100`.
 
 ### Calculating the Weak Subjectivity Period
 
-*Note*: `compute_weak_subjectivity_period()` is planned to be updated when a more accurate calculation is made.
+A detailed analysis of the calculation of the weak subjectivity period is made in [this report](https://github.com/runtimeverification/beacon-chain-verification/blob/master/weak-subjectivity/weak-subjectivity-analysis.pdf). The expressions in the report use fractions, whereas we only use uint64 arithmetic in eth2.0-specs. The expressions have been simplified to avoid computing fractions, and more details can be found [here](https://www.overleaf.com/read/wgjzjdjpvpsd).
+
+#### `get_active_validator_count`
+
+```python
+def get_active_validator_count(state: BeaconState) -> uint64:
+    active_validator_count = len(get_active_validator_indices(state, get_current_epoch(state)))
+    return active_validator_count
+```
+
+#### `compute_avg_active_validator_balance`
+
+```python
+def compute_avg_active_validator_balance(state: BeaconState) -> Ether:
+    total_active_balance = get_total_active_balance(state)
+    active_validator_count = get_active_validator_count(state)
+    avg_active_validator_balance_gwei = total_active_balance // active_validator_count
+    avg_active_validator_balance_eth = avg_active_validator_balance_gwei // ETH_TO_GWEI
+    return avg_active_validator_balance_eth
+```
+
+#### `compute_weak_subjectivity_period`
 
 ```python
 def compute_weak_subjectivity_period(state: BeaconState) -> uint64:
-    weak_subjectivity_period = MIN_VALIDATOR_WITHDRAWABILITY_DELAY
-    validator_count = len(get_active_validator_indices(state, get_current_epoch(state)))
-    if validator_count >= MIN_PER_EPOCH_CHURN_LIMIT * CHURN_LIMIT_QUOTIENT:
-        weak_subjectivity_period += SAFETY_DECAY * CHURN_LIMIT_QUOTIENT // (2 * 100)
+    ws_period = MIN_VALIDATOR_WITHDRAWABILITY_DELAY
+    N = get_active_validator_count(state)
+    t = compute_avg_active_validator_balance(state)
+    T = MAX_EFFECTIVE_BALANCE // 10**9
+    delta = get_validator_churn_limit(state)
+    Delta = MAX_DEPOSITS * SLOTS_PER_EPOCH
+    D = SAFETY_DECAY
+
+    case = (
+        T * (200 + 3 * D) < t * (200 + 12 * D)
```

Comment:
`T * (200 + 3 * D)` and `t * (200 + 12 * D)` are reused. Is there maybe a high level var name that would make these more descriptive?

Reply:
These terms don't correspond to any real-world factor as such, so I think it's best to use them raw in every instance. This also makes it easier to compare the code to the mathematical expressions in the WS report.